### PR TITLE
Veselov Mikhail. Lab 1. Var 4.

### PR DESCRIPTION
**Pull request by Puchiglass**
Implemented identifier renaming (variable name, function name, class name).

### CODE REVIEW

File: clang/labs/lab1/veselov_mikhail/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+add_llvm_library(VeselRenamePlugin MODULE RenamingIdentifier.cpp PLUGIN_TOOL clang)
+
+if(WIN32 OR CYGWIN)
+  set(LLVM_LINK_COMPONENTS
+    Support
+  )
+  clang_target_link_libraries(DeprecatedWarningPlugin PRIVATE
```
Comment (m-ly4) [inline clang/labs/lab1/veselov_mikhail/CMakeLists.txt]: Please fix the name

File: clang/test/lab1/veselov_mikhail/test.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+// NON_EXIST_CLASS-NEXT: A() {};
+// NON_EXIST_CLASS-NEXT: ~A() {};
+// NON_EXIST_CLASS-NEXT: };
+// NON_EXIST_CLASS: void func() {
+// NON_EXIST_CLASS-NEXT: A var1;
+// NON_EXIST_CLASS-NEXT: A* var2 = new A;
+// NON_EXIST_CLASS-NEXT: delete var2;
+// NON_EXIST_CLASS-NEXT: }
+
+// RUN: %clang_cc1 -load %llvmshlibdir/VeselRenamePlugin%pluginext\
```
Comment (m-ly4) [inline clang/test/lab1/veselov_mikhail/test.cpp:188]: Could you please split tests below as well?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/veselov_mikhail/test.cpp (hunk 2 above)
Comment (m-ly4) [reply clang/test/lab1/veselov_mikhail/test.cpp:188]: Sorry, previous comment was improper.
Clarifying: I'd like to see CHECK strings closer to code that they validate.

Comment (m-ly4) [reply clang/test/lab1/veselov_mikhail/test.cpp:188]: `RUN` lines should be written before first split file. Only `CHECK`-like lines may be moved under split test case scope.

File: clang/test/lab1/veselov_mikhail/test.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  a++;
+  return a;
+}
+
+// RUN: %clang_cc1 -load %llvmshlibdir/VeselRenamePlugin%pluginext\
+// RUN: -add-plugin rename\
+// RUN: -plugin-arg-rename type=var\
+// RUN: -plugin-arg-rename cur-name=c\
+// RUN: -plugin-arg-rename new-name=new_var %t/rename_non_existent_var.cpp
+// RUN: FileCheck %s < %t/rename_non_existent_var.cpp --check-prefix=NON_EXIST_VAR
```
Comment (m-ly4) [inline clang/test/lab1/veselov_mikhail/test.cpp]:
```suggestion
// RUN: %clang_cc1 -load %llvmshlibdir/VeselRenamePlugin%pluginext\
// RUN: -add-plugin rename\
// RUN: -plugin-arg-rename type=var\
// RUN: -plugin-arg-rename cur-name=a\
// RUN: -plugin-arg-rename new-name=new_var %t/rename_var.cpp
// RUN: FileCheck %s < %t/rename_var.cpp --check-prefix=VAR

// RUN: %clang_cc1 -load %llvmshlibdir/VeselRenamePlugin%pluginext\
// RUN: -add-plugin rename\
// RUN: -plugin-arg-rename type=var\
// RUN: -plugin-arg-rename cur-name=c\
// RUN: -plugin-arg-rename new-name=new_var %t/rename_non_existent_var.cpp
// RUN: FileCheck %s < %t/rename_non_existent_var.cpp --check-prefix=NON_EXIST_VAR

//--- rename_var.cpp

// VAR: int func() {
// VAR-NEXT: int new_var = 2, b = 2;
// VAR-NEXT: new_var = b + new_var;
// VAR-NEXT: new_var++;
// VAR-NEXT:  return new_var;
// VAR-NEXT: }

int func() {
  int a = 2, b = 2;
  a = b + a;
  a++;
  return a;
}
```

Comment (m-ly4) [reply clang/test/lab1/veselov_mikhail/test.cpp]: This suggestion is provided just as example

File: clang/test/lab1/veselov_mikhail/test.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+//--- rename_non_existent_var.cpp
+
+// NON_EXIST_VAR: int func() {
+// NON_EXIST_VAR-NEXT: int a = 2;
+// NON_EXIST_VAR-NEXT: int b = 3;
```
Comment (m-ly4) [inline clang/test/lab1/veselov_mikhail/test.cpp:37]:
```suggestion
// NON_EXIST_VAR-NEXT: int bbbb = 3;
```
Then let's look if this test fails or not.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/veselov_mikhail/test.cpp (hunk 4 above)
Comment (Puchiglass) [reply clang/test/lab1/veselov_mikhail/test.cpp:37]: Do you mean that you need to rename all variables "b" to "bbbb" in the code we handle with our plugin?

Comment (m-ly4) [reply clang/test/lab1/veselov_mikhail/test.cpp:37]: I meant to validate on your own if those tests are executed at all or not.
You may do that by intentional making an unexpected check that should fail normally. If it isn't then those RUN commands are not executed to some reason.

Comment (Puchiglass) [reply clang/test/lab1/veselov_mikhail/test.cpp:37]: I changed it, it crashes for me, do I need to upload it to the pull request?

Comment (m-ly4) [reply clang/test/lab1/veselov_mikhail/test.cpp:37]: Hmm, OK then. No, do not change the test.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**